cpp/velox/memory/VeloxMemoryPool.cc

```diff
@@ -122,91 +95,126 @@ class VeloxMemoryAllocator final : public velox::memory::MemoryAllocator {
+    }
+    // NOTE: no matter memory pool abort throws or not, it should have been marked
+    // as aborted to prevent any new memory arbitration triggered from the aborted
+    // memory pool.
+    VELOX_CHECK(pool->aborted());
+  }
 
-static std::shared_ptr<velox::memory::MemoryPool> rootVeloxMemoryPool() {
-  static auto options = gluten::VeloxInitializer::get()->getMemoryPoolOptions();
-  static int64_t spillThreshold = gluten::VeloxInitializer::get()->getSpillThreshold();
-  static std::shared_ptr<VeloxMemoryPool> defaultPoolRoot = std::make_shared<VeloxMemoryPool>(
-      nullptr,
-      "root",
-      velox::memory::MemoryPool::Kind::kAggregate,
-      getDefaultVeloxMemoryManager(),
-      facebook::velox::memory::MemoryReclaimer::create(),
-      nullptr,
-      options);
-
-  auto f = [&]() {
-    defaultPoolRoot->setHighUsageCallback(
-        [=](velox::memory::MemoryPool& pool) { return pool.reservedBytes() >= spillThreshold; });
-    defaultPoolRoot->testingSetCapacity(facebook::velox::memory::kMaxMemory);
-    return true;
-  };
+  gluten::AllocationListener* listener_;
+};
 
-  static bool do_once = f();
-  (void)do_once; // suppress warning
+velox::memory::MemoryManager* getDefaultVeloxMemoryManager() {
+  static velox::memory::IMemoryManager::Options mmOptions{};
+  static velox::memory::MemoryManager mm{mmOptions};
+  return &mm;
+}
 
-  return defaultPoolRoot;
+static std::shared_ptr<velox::memory::MemoryPool> rootVeloxMemoryPool() {
+  auto* mm = getDefaultVeloxMemoryManager();
+  static auto root = mm->addRootPool(
+      "gluten_root", mm->capacity(), facebook::velox::memory::MemoryReclaimer::create()); // the 3rd capacity
+  return root;
 }
 
 std::shared_ptr<velox::memory::MemoryPool> defaultLeafVeloxMemoryPool() {
-  static std::shared_ptr<velox::memory::MemoryPool> defaultPool = rootVeloxMemoryPool()->addLeafChild("default_leaf");
-  return defaultPool;
+  // this pool is not tracked by Spark
+  static auto leaf =
+      rootVeloxMemoryPool()->addLeafChild("default_leaf", true, facebook::velox::memory::MemoryReclaimer::create());
+  return leaf;
 }
 
 std::shared_ptr<velox::memory::MemoryPool> asAggregateVeloxMemoryPool(gluten::MemoryAllocator* allocator) {
+  // this pool is tracked by Spark
   static std::atomic_uint32_t id = 0;
-  auto pool = rootVeloxMemoryPool()->addAggregateChild(
-      "wrapped_root" + std::to_string(id++), facebook::velox::memory::MemoryReclaimer::create());
-  auto wrapped = std::dynamic_pointer_cast<VeloxMemoryPool>(pool);
-  VELOX_CHECK_NOT_NULL(wrapped);
-  velox::memory::MemoryAllocator* veloxAlloc = wrapped->getAllocator();
-  wrapped->setAllocatorShared(std::make_shared<VeloxMemoryAllocator>(allocator, veloxAlloc));
+  velox::memory::MemoryAllocator* veloxAlloc = velox::memory::MemoryAllocator::getInstance();
+  gluten::MemoryAllocator* glutenAlloc;
+  gluten::AllocationListener* listener;
+  if (dynamic_cast<gluten::ListenableMemoryAllocator*>(allocator)) {
```

Comment:
consider adding an interface `type()` for `gluten::MemoryAllocator` to tell `what is it`

Reply:
Did you mean a `enum` or `std::string` type?

Reply:
yes